Fix cylc clean while cat-log running

datamel opened the pull request:
`cylc clean` will fail when a cat-log process is keeping one of the log files open. This is a more pressing issue now that the log view is available in the gui.
This is related to an nfs issue and may not be possible to reproduce on all systems.

The solution involves 
- changing the cat-log command to include `--follow=name`, changed in the default.  
- changing `rmtree` behaviour, by retrying on error.

closes https://github.com/cylc/cylc-flow/issues/5341 

Note for debugging this issue, I changed the error message to be passed into the gui.

To test, open the gui and open a log file. With the log view still open, clean the workflow.
This will error on master but not on this branch.

In the gui, I think we should maybe close any open views on clean, perhaps even redirect to the dashboard? But this would be a separate issue.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [ ] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

oliver-sanders commented:
> If this is a bug fix, PR should be raised against the relevant ?.?.x branch.

Needs rebasing onto 8.1.x.

MetRonnie commented:
> Note for debugging this issue, I changed the error message to be passed into the gui, so this (if reviewers are happy with the approach) will also close #5324.

Thanks, however this isn't sufficient to close that issue. We need to check for other instances of logging that should be moved into the exception message

datamel commented:
> > Note for debugging this issue, I changed the error message to be passed into the gui, so this (if reviewers are happy with the approach) will also close #5324.
> 
> Thanks, however this partially addresses that issue. We need to check for other instances of logging that should be moved into the exception message

Okay, I've removed the close from the description.

oliver-sanders commented on cylc/flow/pathutil.py:
```diff
@@ -297,15 +298,27 @@ def remove_dir_and_target(path: Union[Path, str]) -> None:
+        try:
+            rmtree(target)
+            return
+        except OSError as exc:
```
To avoid retrying the `shutil.rmdirs` when it is not appropriate to do so we should filter for "directory not empty" errors. The `errno` module is the tool for the job (error numbers are OS dependent):

```python
        except OSError as exc:
            if exc.errno == errno.ENOTEMPTY:
                sleep(sleep_time)
            else:
                raise
```

https://docs.python.org/3/library/errno.html?highlight=errno#errno.ENOTEMPTY

MetRonnie replied:
That's how `handle_rmtree_err` works

MetRonnie replied:
Why the decision to retry 5 times? (There is no docstring for this function)

MetRonnie commented on cylc/flow/scripts/clean.py:
```diff
@@ -193,15 +193,18 @@ async def run(*ids: str, opts: 'Values') -> None:
         except Exception as exc:
-            failed.append(workflow)
+            failed[workflow] = exc
             LOG.warning(exc)
```
I don't think we'll need the warning as its a duplicate of the error below
```suggestion
```

MetRonnie commented:
Ah, just realised a slight flaw with only testing for `ENOTEMPTY` - the second time it tries it, the error will be `EBUSY` instead, so it will only retry once before failing. (Assuming it's a "sillyrenamed" `.nfsXXXX...` file causing the problem)

![image](https://user-images.githubusercontent.com/61982285/225889400-c9cc07d8-2425-463b-8328-c3c7678b2a32.png)

Also annoying is how `rmtree` only gives the filename, not path. If it's a job log file that's being kept open by a running job, there could be hundreds of job log dirs and you don't know which one contains the `.nfsXXXX...`.

Anyway, we can sort these out in a follow-up issue.

oliver-sanders commented:
> Ah, just realised a slight flaw with only testing for ENOTEMPTY - the second time it tries it, the error will be EBUSY instead

Dammit, confirmed with this script:

```bash
#!/usr/bin/env bash                                                      
                                                                         
set -eux                                  
                                          
TD="$HOME/cylc-run/foo/log/scheduler"    
mkdir -p "$TD"         
TF="$TD/log"           
                       
echo -e 'a\nb\nc' > "$TF"    
tree "$TD"             
                       
tail --follow=name -f "$TF" & PROC="$!"    
     
sleep 2    
     
# this will likely fail due to FS locking stuffs    
rm -r "$TD" && echo 'REMOVE PASSED' || echo 'REMOVE FAILED'    
rm -r "$TD" && echo 'REMOVE PASSED' || echo 'REMOVE FAILED'    
rm -r "$TD" && echo 'REMOVE PASSED' || echo 'REMOVE FAILED'                                                                        
                                             
# tidy up after                              
kill "$PROC"                                 
rm -r "$TD" || true     
```

Which reliably gives:

* Directory not empty
* Device or resource busy
* Device or resource busy

oliver-sanders commented on cylc/flow/pathutil.py:
```diff
@@ -297,15 +299,35 @@ def remove_dir_and_target(path: Union[Path, str]) -> None:
+            rmtree(target)
+            return
+        except OSError as exc:
+            if exc.errno == errno.ENOTEMPTY:
```
```suggestion
            if exc.errno in {errno.ENOTEMPTY, errno.EBUSY}:
```

oliver-sanders commented:
I used this script to double-check the premise that the tail process dies when the file is removed and to confirm the error message(s) returned:

```bash
#!/usr/bin/env bash                                                      
                                                                         
set -eu                                                            
                                                 
TD="$HOME/cylc-run/foo/log/scheduler"            
mkdir -p "$TD"                                   
TF="$TD/log"                                     
                                                 
echo -e 'a\nb\nc' > "$TF"                        
                                                 
# NOTE: for some reason the tail command doesn't die when launched
# as a background process                        
echo 'Start you tail process now:'               
echo "i.e. tail -f --follow=name '$TF'"          
sleep 5                                          
                                                 
# retry removal a few times                      
SLEEP=1                                          
for try in $(seq 1 10); do                       
    sleep "$SLEEP"                               
    echo "Try: $try"                             
    rm -r "$TD" || continue                      
    echo "Success after $((try * SLEEP)) seconds."
    break            
done                 
                     
# tidy up after JIC
rm -r "$TD" 2>/dev/null || true 
```

This is hard to test, would need NFS in a container and whatnot so commenting it here for the record, will link this comment into the docstring.

oliver-sanders commented on cylc/flow/pathutil.py:
```diff
@@ -297,15 +299,35 @@ def remove_dir_and_target(path: Union[Path, str]) -> None:
+def _rmtree(target: Union[Path, str],
+            retries: int = 5,
+            sleep_time: float = 0.5):
+    """Make rmtree more robust to nfs issues.
+
+    shutil.rmtree will be retried (default 5 times)
```
```suggestion
    """Make rmtree more robust to nfs issues.
    
    If a file is deleted which is being held open for reading by another process
    NFS will create a ".nfs" file in the containing directory to handle this.

    If you try to delete the directory which contains these files you will get either
    a ENOTEMPTY or EBUSY error.
    
    A likely cause of open file handles in cylc-run directories is `cylc cat-log -m t`.
    If the file being cat-log'ged is removed, the command will fail on its next poll.
    The default poll interval is one second, so if we wait a couple of seconds and
    retry the removal it will likely work.

    This command retries removal a specified number of times at a specified
    interval before failing to give cat-log process a chance to die gracefully and
    release their filesystem locks. For more info see:
    https://github.com/cylc/cylc-flow/pull/5359#issuecomment-1479989975
```

oliver-sanders commented on cylc/flow/pathutil.py:
```diff
@@ -297,15 +299,49 @@ def remove_dir_and_target(path: Union[Path, str]) -> None:
+        _rmtree(path)
+
+
+def _rmtree(target: Union[Path, str],
+            retries: int = 5,
+            sleep_time: float = 0.5):
+    """Make rmtree more robust to nfs issues.
+    
+    If a file is deleted which is being held open for reading by another process
+    NFS will create a ".nfs" file in the containing directory to handle this.
+
+    If you try to delete the directory which contains these files you will get either
+    a ENOTEMPTY or EBUSY error.
+    
+    A likely cause of open file handles in cylc-run directories is `cylc cat-log -m t`.
+    If the file being cat-log'ged is removed, the command will fail on its next poll.
+    The default poll interval is one second, so if we wait a couple of seconds and
+    retry the removal it will likely work.
+
+    This command retries removal a specified number of times at a specified
+    interval before failing to give cat-log process a chance to die gracefully and
+    release their filesystem locks. For more info see:
+    https://github.com/cylc/cylc-flow/pull/5359#issuecomment-1479989975
+    """
```
```suggestion
def _rmtree(
    target: Union[Path, str],
    retries: int = 5,
    sleep_time: float = 0.5,
):
    """Make rmtree more robust to nfs issues.

    If a file is deleted which is being held open for reading by
    another process. NFS will create a ".nfs" file in the
    containing directory to handle this.

    If you try to delete the directory which contains these
    files you will get either a ENOTEMPTY or EBUSY error.

    A likely cause of open file handles in cylc-run directories
    is `cylc cat-log -m t`. If the file being cat-log'ged is removed,
    the command will fail on its next poll. The default poll
    interval is one second, so if we wait a couple of seconds and
    retry the removal it will likely work.

    This command retries removal a specified number
    of times at a specified interval before failing to
    give cat-log process a chance to die gracefully and
    release their filesystem locks. For more info see:
    https://github.com/cylc/cylc-flow/pull/5359#issuecomment-1479989975
    """
```

oliver-sanders commented:
Found a `rmdir` call which might need to be converted to `_rmtree`?

```diff
diff --git a/cylc/flow/pathutil.py b/cylc/flow/pathutil.py
index 3acde3700..173a82753 100644
--- a/cylc/flow/pathutil.py
+++ b/cylc/flow/pathutil.py
@@ -397,8 +397,8 @@ def remove_empty_parents(
         if not parent.is_dir():
             continue
         try:
-            parent.rmdir()
             LOG.info(f'Removing directory: {parent}')
+            _rmtree(path)
         except OSError:
             break
```

MetRonnie reviewed (approved):
I've tested it out. Added a changelog entry
